src/mlpro/bf/math/normalizers.py

```diff
@@ -0,0 +1,319 @@
+        """
+        method to normalize element
+
+        Parameters
+        ----------
+        p_element:Element or a numpy array
+            Element to be normalized
+
+        Returns
+        -------
+        element:Element or numpy array
+            Normalized element
+        """
+
+        if self._param is None:
+            raise ImplementationError('normalization parameters not set')
+        # element = self._normalize(p_element, p_param = self._param)
+        p_param = self._param
+        if isinstance(p_element, Element):
+            normalized_element = Element(p_element.get_related_set())
+            normalized_element.set_values(np.multiply(p_element.get_values(), p_param[0]) - p_param[1])
+        elif isinstance(p_element, np.ndarray):
+            normalized_element = np.multiply(p_element, p_param[0]) - p_param[1]
+        else: raise ParamError('wrong data type provided for normalization')
+        return normalized_element
+
+
+## -------------------------------------------------------------------------------------------------
+    def denormalize(self, p_element:Union[Element, np.ndarray], p_param=None):
+        """
+        Method to denormalize the normalized elements.
+
+        Parameters
+        ----------
+        p_element:Element or a numpy array
+            Element to be denormalized
+        p_param  -Optional
+            Parameters to be normalized
+
+        Returns
+        -------
+        element:Element or numpy array
+            Denormalized element
+        """
+        if self._param is None:
+            raise ImplementationError('normalization parameters not set')
+        if p_param is None:
+            p_param = self._param
+        if isinstance(p_element, Element):
+            denormalized_element = Element(p_element.get_related_set())
+            denormalized_element.set_values(np.multiply(p_element.get_values(), 1 / p_param[0]) + (
+                        p_param[1] / p_param[0]))
+        elif isinstance(p_element, np.ndarray):
+            denormalized_element = np.multiply(p_element, 1 / p_param[0]) + \
+                                   (p_param[1] / p_param[0])
+        else:
+            raise ParamError('wrong datatype provided for denormalization')
+        return denormalized_element
+
+
+## -------------------------------------------------------------------------------------------------
+    def renormalize(self, p_data:Union[Element, np.ndarray]):
+        """
+        Method to denormalize and renormalize an element based on old and current normalization parameters.
+
+        Parameters
+        ----------
+        p_element:Element or numpy array
+            Element to be renormalized.
+
+        Returns
+        -------
+        renormalized_elemet:Element or numpy array
+            Renormalized element
+
+        """
+
+        denormalized_element = self.denormalize(p_data, p_param = self._param_old)
```

Comment:
Please use method self._set_parameters() to switch between old/current parameter sets.